packages/instantsearch.js/src/lib/insights/listener.tsx

```diff
@@ -2,23 +2,58 @@
 
 import { h } from 'preact';
 
-import { readDataAttributes, hasDataAttributes } from '../../helpers/insights';
-import { deserializePayload } from '../utils';
+import { readDataAttributes } from '../../helpers/insights';
+import { deserializePayload, warning } from '../utils';
 
 import type { InsightsEvent } from '../../middlewares/createInsightsMiddleware';
 import type { InsightsClient } from '../../types';
 
-type WithInsightsListenerProps = {
-  [key: string]: unknown;
-  insights: InsightsClient;
-  sendEvent?: (event: InsightsEvent) => void;
+export type InsightsEventHandlerOptions = {
+  insights?: InsightsClient;
+  sendEvent: (event: InsightsEvent) => void;
 };
 
-const findInsightsTarget = (
+export const createInsightsEventHandler =
+  ({ insights, sendEvent }: InsightsEventHandlerOptions) =>
+  (event: MouseEvent): boolean => {
+    let hasSentAnEvent = false;
+
+    // new way, e.g. bindEvent("click", hit, "Hit clicked")
+    const insightsThroughSendEvent = findInsightsTarget(
+      event.target as HTMLElement | null,
+      event.currentTarget as HTMLElement | null,
+      (element) => element.hasAttribute('data-insights-event')
+    );
+
+    if (insightsThroughSendEvent) {
+      const payload = parseInsightsEvent(insightsThroughSendEvent);
+
+      payload.forEach((single) => sendEvent(single));
+      hasSentAnEvent = true;
+    }
+
+    // old way, e.g. instantsearch.insights("clickedObjectIDsAfterSearch", { .. })
+    const insightsThroughFunction = findInsightsTarget(
+      event.target as HTMLElement | null,
+      event.currentTarget as HTMLElement | null,
+      (element) =>
+        element.hasAttribute('data-insights-method') &&
+        element.hasAttribute('data-insights-payload')
+    );
+    if (insightsThroughFunction) {
+      const { method, payload } = readDataAttributes(insightsThroughFunction);
+      insights!(method, payload);
+      hasSentAnEvent = true;
+    }
+
+    return hasSentAnEvent;
```

Comment:
I'm not sure to understand why we need to return a boolean from the handler?

Reply:
it's not really needed for this PR indeed, it was with #5527 in mind. I'll revert here